wptherml/optdriver.py

```diff
@@ -0,0 +1,1205 @@
+import numpy as np
+from scipy.optimize import minimize
+from .em import TmmDriver
+import torch
+import torch.nn as nn
+import torch.optim as optim
+import tqdm
+from sklearn.model_selection import train_test_split
+from sklearn.preprocessing import StandardScaler
+from sklearn.preprocessing import MinMaxScaler
+import copy
+from pyqubo import Binary
+import neal
+import pandas as pd
+
+class OptDriver(TmmDriver):
+    """Compute the absorption, scattering, and extinction spectra of a sphere using Mie theory
+
+    Attributes
+    ----------
+    TBD
+
+    Returns
+    -------
+    None
+
+    Examples
+    --------
+    >>> fill_in_with_actual_example!
+    """
+
+
+    def __init__(self, args):
+        #store args
+        self.args = args
+
+        # ordinary tmm driver input
+        args = {k.lower(): v for k, v in args.items()}
+        self.parse_input(args)
+        self.parse_optimization_input(args)
+        self.set_refractive_index_array()
+        # compute reflectivity spectrum
+        self.compute_spectrum()
+        #print("We started optimizing")
+        #self.optimize()
+
+
+    def parse_optimization_input(self, args):
+        """ Parse additional options related to the optimization, including:
+            - the objective function to be optimized 
+            - minimization or maximization of the objective
+            - the optimization method to be used
+            - lower- and upper-bounds on the layer thicknesses
+            
+        """
+        # which objective function do we want to optimize
+        if "objective_function" in args:
+            self.objective_function = args["objective_function"]
+        # default to selective mirror    
+        else:
+            self.objective_function = "selective_mirror"
+
+        # do we want to minimze the objective?
+        if "minimization" in args:
+            self.mimimization = args["minimization"]
+        # usually we want to mazimize, so default is false
+        else: 
+            self.minimization = False
+        
+        # set bounds on thickness of each layer
+        if "lower_bound" in args:
+            self.lower_bound = args["lower_bound"]
+        else:
+            self.lower_bound = 1 
+
+        if "upper_bound" in args:
+            self.upper_bound = args["upper_bound"]
+        else:
+            self.upper_bound = 1000
+
+
+        if "combinatorial_optimization" in args:
+            self.combinatorial_optimization = True
+            self.qubo_thickness_optimization = False
+            self.optimization_dictionary = args['optimization_dictionary']
```

Comment:
@PeytonRoden I have a few thoughts about optimization_dictionary: (1) can the entries of this dictionary just fit into the dictionary that the user passes as args?  (2) we need to figure an easy and comprehensive way to provide documentation for the user arguments because they are numerous now.  In principle Sphinx can provide documentation from docstrings that could be included in the parse_input(), parse_optimization_input(), and parse_therml_input() methods... that might be the way to go, we should look into this.